viewer-prototype/src/browser/trace-viewer/components/table-output-component.tsx

```diff
@@ -2,78 +2,178 @@
+                    const item = rowsThisPage[i];
+                    const itemCopy = cloneDeep(item);
+                    rowsThisPage[i] = itemCopy;
+                }
+
+                if (this.props.cacheBlockSize && (rowsThisPage.length < this.props.cacheBlockSize)) {
+                    params.successCallback(rowsThisPage, params.startRow + rowsThisPage.length);
+                } else {
+                    params.successCallback(rowsThisPage);
+                }
+            }
+        };
+        event.api.setDatasource(dataSource);
+    };
+
+    private async init() {
         const traceUUID = this.props.traceId;
         const tspClient = this.props.tspClient;
         const outPutId = this.props.outputDescriptor.id;
 
         // Fetch columns
         const columnsResponse = (await tspClient.fetchTableColumns<Entry, EntryHeader>(traceUUID, outPutId, QueryHelper.timeQuery([0, 1]))).getModel();
         const columnEntries = columnsResponse.model.entries;
-        const columnIds: Array<number> = [];
+        const colIds: Array<number> = [];
         const columnsArray = new Array<any>();
+
+        if (this.showIndexColumn) {
+            columnsArray.push({
+                headerName: 'Index',
+                field: '0',
+                width: this.props.columnWidth,
+                cellRenderer: 'loadingRenderer'
+            });
+            colIds.push(0);
+        }
+
         columnEntries.forEach(entry => {
-            columnIds.push(entry.id);
+            const id = this.showIndexColumn ? ++entry.id : entry.id;
+            colIds.push(id);
             let columnName = '';
             if (entry.labels.length) {
                 columnName = entry.labels[0];
             }
             columnsArray.push({
                 headerName: columnName,
                 field: entry.id.toString(),
-                width: 200
+                width: this.props.columnWidth
             });
         });
 
-        // Fetch lines
-        const lineResponse = (await tspClient.fetchTableLines(traceUUID, outPutId, QueryHelper.tableQuery(columnIds, 0, 500))).getModel();
-        const model = lineResponse.model;
-        const lines = model.lines;
-        const linesArray = new Array<any>();
-        lines.forEach(line => {
-            const obj: any = {};
-            const cells = line.cells;
-            const ids = model.columnIds;
-            for (let i = 0; i < cells.length; i++) {
-                obj[ids[i]] = cells[i].content;
-            }
-            linesArray.push(obj);
-        });
+        if (!this.showIndexColumn) {
+            columnsArray[0].cellRenderer = 'loadingRenderer';
+        }
 
-        this.setState({
-            outputStatus: lineResponse.status,
-            tableColumns: columnsArray,
-            tableLines: linesArray
-        });
+        this.columnIds = colIds;
+        this.columnArray = columnsArray;
+
+        this.state = {
```

Comment:
Don't set the state directly, use this.setState instead.

Reply:
Done.